### ISSUE TITLE
Traffic Manager Integration

### ISSUE THREAD

**Issue by lindydonna** [Azure/Azure-Functions#214]
_From @arafato on May 23, 2016 19:4_

Hi--, 

is integration of Azure Functions with Traffic Manager (TM) supposed to work? I've just tried it out, however, TM is not able to resolve the endpoints. Thanks!

Best regards
--Oliver.

_Copied from original issue: Azure/azure-webjobs-sdk-script#379_

**Comment by lindydonna** [comment]
_From @davidebbo on May 23, 2016 19:21_

At this time, Function Apps in the Dynamic sku have a limitation that they don't support custom host names. It lets you register them but they don't actually work, which I think is why TM is not working.

This is something that we're planning to enable, but will take a few iteration. That logic sits outside of the GitHub code base, but it's ok to track it here.

**Comment by lindydonna** [comment]
_From @christopheranderson on May 24, 2016 22:41_

Assigning to myself to track

**Comment by lindydonna** [comment]
_From @christopheranderson on June 21, 2016 20:32_

Need to chat with platform folks to see if this will be around for rc. Might change milestone to help me track.

**Comment by lindydonna** [comment]
_From @DarinMacRae on September 26, 2016 18:12_

Just ran into this one myself. Any status update on this? I'm using ~0.6.

**Comment by lindydonna** [comment]
_From @AndyMDoyle on November 16, 2016 8:21_

Any update on this now Azure Functions has hit release?

**Comment by lindydonna** [comment]
_From @DKincer on March 16, 2017 13:5_

*bump* Checking if there is an update on this issue. Currently publishing an Azure Function takes it offline. Need to set up a priority traffic manager to divert from East to West during a publish

**Comment by lindydonna** [comment]
@DKincer We'll be enabling this feature soon in the UI. I'll see if there's a workaround using ARM APIs in the meantime.

**Comment by lindydonna** [comment]
@DKincer Traffic Manager does not work. I'm keeping this issue open to track the feature request.

**Comment by DKincer** [comment]
Thank you.

**Comment by CANUSA** [comment]
Do we have any type of ETA on this particular feature? I could really use it!

**Comment by rawlukMike** [comment]
It would be nice to add some information about this. Currently traffic manager is working for some time when adding as external. BUT even if I am adding: http://<functions-name>.azurewebsites.net,
Traffic manager actually points to http://HOSTS.<functions-name>.azurewebsites.net, which pulls up azure websites 404, and does not work with functions:

> nslookup dummiertf.trafficmanager.net
> 
> Non-authoritative answer:
> Name:    hosts.emotiondemo.azurewebsites.net
> Address:  52.169.184.163
> Aliases:  dummiertf.trafficmanager.net
>           emotiondemo.azurewebsites.net

**Comment by rawlukMike** [comment]
According to:
[App service SKU comparison matrix](https://blogs.msdn.microsoft.com/appserviceteam/2017/03/23/azure-app-service-web-api-mobile-ase-azure-functions-sku-comparison-matrix-2/)
Traffic Manager Integration SHOULD work for functions cause it's ok with consumption plan. 
And it does... for a while.
Then traffic manager goes to STOPPED state for functions and its no longer working.
There is NO DOCUMENTATION for every step necessary.
To set up traffic manager i do this:
```
Create a new External Endpoint (as suggested  in doc linked)
Route /api/ping to one of functions (should be stated in docs)
Add trafficmanager as custom hostname (should be stated in docs)
```
After this steps traffic manager pick up function and makes it ONLINE...
But after some time it gets STOPPED and won't work.
I contacted traffic manager support, since it seems like that was their mistake, but they say that it's not their fault and point out that contacting *.azurewebsites.net trigger another redirection to hosts.azurewebsites.net.

Can you investigate this ? This is clearly a bug on either traffic manager, or on functions redirection from hosts.**.azurewebsites.net.

Functions would be such a great tool if integration with traffic manager worked as expected.
If you cannot fix it then REMOVE MISSLEADING information from docs above.

Thanks for reading :)

**Comment by LarsKemmann** [comment]
@lindydonna Can you comment on @rawlukMike's reported issue above? We're evaluating a geo-distributed Functions app as well.

**Comment by lindydonna** [comment]
@rawlukMike This sounds like a limitation of external endpoints. My guess is that when the Function App goes idle, it ends up going offline in Traffic Manager. So, your function would need to be running continuously for this to work.

We'll investigate and update the blog post with additional information.

**Comment by rawlukMike** [comment]
@lindydonna Sorry I shall update my answer above tomorrow.
I solved the issue by investigation step by step.
What's actually solving all problems with traffic manager is creating a PING function (the best is javascript since its fastest) that respond to functionName.azurewebsites.net/api/ping.
Only if you have custom function that respond on ping, external endpoint won't die after a while.
This is actually pretty obvious, but even TM expert from premium support could not solve it.
This is not nicest solution since it generates cost since traffic manager is executing function constantly but still within free limits, so not that costly.

I will update my post above so it will have solution, but you shall point this in docs about websites that this is required.

**Comment by sympthom** [comment]
@rawlukMike 

I I tried your 2-steps, however there must be some missing steps in order to have a fully functional solution. The probe in TM is now working (probing the Function App) and will there for create the alias for the name of the Function App. However the Function App doesn't resolve. How to configure the host name in the Function App for the new TM-alias?

**Comment by rawlukMike** [comment]
This is what we do in our solution:
Set up ping function in javascript (fastest) so /api/ping will respond
nicely to TM.
Your domain must point to TM.
Add external endpoint to functions.
Change priority so TM points to to functions <your domain should follow so
its downtime start>
Assuming you did ok your domain MUST point to functions through TM.
Now add custom hostname in functions with YOURDOMAIN (not tm)
Azure will check if your domain hits functions (yes it do through tm), and
add new hostname.
Now you can redirect traffic in TM.
Hostname stays.

2017-10-10 20:53 GMT+02:00 Martin Dahlstedt <notifications@github.com>:

**Comment by sympthom** [comment]
@rawlukMike 

Thanks for great feedback! Is there a way to create this domain/CNAME in Azure in order to point it to TM and also to be added as a custom hostname in Azure Function?

**Comment by rawlukMike** [comment]
Yes, you can buy it in Azure.
https://docs.microsoft.com/en-us/azure/app-service/custom-dns-web-site-buydomains-web-app

But before you post anymore here, please consider if it's related to topic
:).


2017-10-10 22:54 GMT+02:00 Martin Dahlstedt <notifications@github.com>:

**Comment by HansOlavS** [comment]
Using TM on consumption plan App Services now works fine. I just tested it with SSL and everything! All checks out fine. This issue should be closed. @AzureFunctions told me a fix was rolled out yesterday on their twitter handle (https://twitter.com/AzureFunctions/status/925483053677346817)

**Comment by paulbatum** [comment]
Indeed, this is fixed. Closing.

**Comment by rawlukMike** [comment]
THIS IS NOT FIXED.
Create function with route: /ping
And Traffic Manager endpoint will go to stopped state.
Far from great solution, since there is no information whatsoever that having name.azurewebsites.net/api/ping will blow away TM integration.

**Comment by sympthom** [comment]
Please stop closing issues that aren't resolved yet.

**Comment by joshdcar** [comment]
I just went through and created a new Azure Function w/ consumption plan, created a ping function that returns OK and allows anonymous requests, created a traffic manager and added the function as an endpoint and update the Endpoint Monitoring path to 'api/ping' in the configuration.   So far everything is working as expected for me.

![azure-functions-traffic-manager](https://user-images.githubusercontent.com/24570240/32327666-c0c6d8f8-bfad-11e7-9dda-adebfeb2c02a.png)

A couple quick things to check if your in a stopped state:
-  Update your Endpoint Monitoring Path to use /api/ping
-  Ensure your 'ping' function is set for anonymous and always returns back a 200 (ok) 

Traffic manager requires an endpoint to monitor that always returns back 200 so I typically add a anonymous "ping" endpoint to all my APIs that i'm using with traffic manager. Otherwise you will quickly get into a stopped state if your endpoint doesn't exist or requires some sort of authentication . Alternatively I suspect you could add the ?code={secret} to the Endpoint Monitoring Path but I haven't tried that.  You can use any authentication scheme need (function, AD, etc) on your actual http endpoints (triggers). 

Hope this helps!

**Comment by rawlukMike** [comment]
@joshdcar There is NO difference to have it as External Endpoint then.
Anyways we have to provide our OWN Ping function responding to /api/ping (or anything we set in TM config), and our logs will be filled with tons of calls from TM. Not to mention we will be charged for calls from TM at some point?

This is NOT integration. This is just manual hack, that require user to create function, set routing and deal with extra executions.

**Comment by joshdcar** [comment]
Hi @rawlukMike - I don't work for Microsoft so I'm unsure of the internals here I was just sharing that it was working for my use case - at least on my initial passes :)  I just discovered this yesterday so I haven't gone fairly deep just a quick "hello world".  For me at the moment I'm in an East Coast (North America) region and it's been up for a couple hours without problems (so far lol).   

My use of custom endpoints was based on my experience working with Traffic Manager on API Apps that required authentication. At the time (and may still be the case) Traffic Manager couldn't probe endpoints behind authentication walls so I had needed that custom endpoint.  I had not noticed the tweet from @HansOlavS about it working without a custom monitored path so thanks for sharing that - it's quite possible that the steps I used are not necessary or alternatively could also be why I'm not stopped yet.  I'll have to test some different variations this afternoon. 

You bring up good points on costs for those pings and the impact on keeping the app alive just for those pings but I'll let the Azure Functions team chime in.  I've been waiting for this feature for a while so here's hoping it can be sorted out :)

**Comment by HansOlavS** [comment]
Hi @joshdcar, I've done exactly the same as you with "api/ping" with anonymous access and it's been running for over a day without any issues. I'm actually in PRODUCTION with this, so I really really hope this feature is as buggy as it seems from your end...

I'm in West Europe/North Europe with my traffic manager endpoints.

**Comment by HansOlavS** [comment]
Hi @joshdcar, I've done exactly the same as you with "api/ping" with anonymous access and it's been running for over a day without any issues. I'm actually in PRODUCTION with this, so I really really hope this feature isn't as buggy as it seems from your end... Can anyone from MS comment on this ASAP?

I'm in West Europe/North Europe with my traffic manager endpoints.

**Comment by HansOlavS** [comment]
@rawlukMike I'm running on endpoints in West Europe + North Europe and can't see any issues. FYI

**Comment by joshdcar** [comment]
@HansOlavS are you consistently getting hits for the last day (e.g. no cold starts)? I ask because I was hitting my endpoint regularly for about two hours to generate some traffic with zero issues.  I went away for a bit to actually get some work done and I came back after 30-45+ minutes and I was in a stopped state.  Not sure if it's due to cold starts (which theoretically should only be the first hit) or there was no running instance - not sure but just more info for the team to help diagnose.

**Comment by rawlukMike** [comment]
I came with Proxies solution.
Instead using separate function just to respond to probe, I override api/ping route response in proxies.json. There is ZERO functions in service just single proxy.

> 
> {
> $schema: http://json.schemastore.org/proxies,
> -proxies: {
> -Ping: {
> -matchCondition: {
> route: "proxy/ping"
> },
> -responseOverrides: {
> response.statusCode: "200",
> response.statusReason: "Fine",
> response.body: "Hello Probe"
> }
> }
> }
> }
> 

Actually whats done in background is Proxy calling something called Functions.Ping (not part of my service).

> 2017-11-02T15:24:34.640 Executing request via Azure Function Proxies
> 2017-11-02T15:24:34.640 Function started (Id=5af21268-d2b3-4616-a8e4-aa9c89f04677)
> 2017-11-02T15:24:34.640 Executing 'Functions.Ping'


No matter if I setup JSFunction or this proxy, I have functions responding properly to URL thats in TM config.
HOWEVER if endpoint is Azure then it will became Stopped after a while.
No problem if Endpoint is External.

Apart from this I state other problem once more.
There are executions, which we are seeing in logs. They also cost us money (nothing above free usage, but still). Also it requires user action to have integration done.
This is not how I image nice user experience.

**Comment by lkems** [comment]
The issue should be fixed by now after we enable registering Function Apps to Traffic Manager.

**Comment by odvoskin** [comment]
Closing the issue per solution mentioned by @lkems. Please let us know if you encounter further issues.

Thanks,
Oded

**Comment by joetechbob** [comment]
I'm still hitting this issue. Initial integration works, but after, e.g., publishing an update to my Azure functions, Traffic Manager ends up in a stopped state and DNS lookup fails. I then need to manually fix up TM to get it going again. This is happening in both uswest and useast.

Feature/fix is not complete.

**Comment by odvoskin** [comment]
@joetechbob - do you mind sharing your subscription ID and app name? You could share these privately to my email and we'll look into this for you. odvoskin@microsoft.com

**Comment by odvoskin** [comment]
@joetechbob - are you still hitting an issue? If so, please follow up on my previous message. Thanks.

**Comment by Crazedpeanut** [comment]
Is the fix being deployed globally? This is still an issue for me.

**Comment by odvoskin** [comment]
@Crazedpeanut - globally...

I'm re-opening the issue after feedback @Crazedpeanut and @joetechbob. The team will be investigating and we'll reply here.

**Comment by lkems** [comment]
We believe the apps are stopped by API requests from a Slave stamp, which is different from the one hosting the app. A fix was checked in to ignore the requests from Slave stamps. The change will be roll out to production in about two weeks. We will know if it resolves the issue then.

**Comment by BadgerCode** [comment]
@lkems Is there any update on this fix?
I tried setting up a HTTP-triggered function on a consumption plan recently and after a short while the traffic manager endpoint is "Stopped", despite the function still being available.

I even have a cron job hitting the function every 5 minutes to keep it warm.

**Comment by odvoskin** [comment]
The fix is actually rolling out in the deployment we have going on right now. Once complete, we intend to validate in prod before updating.

**Comment by odvoskin** [comment]
The fix was rolled out and validated. Please reach out for any issues encountered.

**Comment by tknerr** [comment]
I'm desperately trying to get a Function (as an external endpoint) to work behind a Traffic Manager. The function is only a proxy and returns a static response. TrafficManager shows that the function is "Online". The function app alone is working, but whenever I go via the TrafficManager I get the 404 page (presumably because the last CNAME record in the chain points to "hosts.<functionname>.azurewebsites.net" instead of "functionname.azurewebsites.net", as @rawlukMike mentioned earlier).

I still have no idea how to resolve that issue, but I suspect I must be doing something fundamentally wrong given that all of you are past that problem already.

I have tried a custom `/api/ping` endpoint (via function proxy) and adjusted the monitoring settings in the TM accordingly. I have tried to use a Standard AppService plan instead of a Consumption based plan but did not help either. I have added custom domain name to the function (with the Consumption based plan) but did not help either.

In all of the cases I tried the TM reported the external endpoint (i.e. the function) as "Online", but when I access it via the TM URL I got the famous 404 page. When accessed directly via "<functionname>.azurewebsites.net" it worked.

Any clues what I might be doing wrong?

(this as in Western Europe region btw)
